### PR TITLE
feat(bench-gate): #433 vocab_bridge NDCG@k uplift driver — evidence: bridge regresses on labelled corpus

### PR DESCRIPTION
**Pull request by robotrocketscience**
Adds the strict NDCG@k uplift driver for #433 Phase 2 (the operator-side ship gate). Three atomic commits:

1. `feat(test):` `run_vocab_bridge_uplift` driver in `tests/retrieve_uplift_runner.py` — mirrors `run_doc_linker_uplift` / `run_query_strategy_uplift` / `run_clustering_uplift` shape. Result type `VocabBridgeUplift{n_rows, mean_ndcg_off, mean_ndcg_on}` with `.uplift` property.
2. `test(uplift):` three unit tests (empty input, skip-without-`expected_top_k`, synthetic-row smoke).
3. `test(bench-gate):` `test_vocab_bridge_ship_gate_runner_present` in `tests/bench_gate/test_vocab_bridge_uplift.py` — same skip-if-runner-absent + skip-if-no-`expected_top_k` pattern as #436. Module docstring updated for the extended row schema (`expected_top_k` joins `expected_canonicals`).

## Bench evidence

I authored a 20-row labelled corpus on the lab side (`~/projects/aelfrice-lab/tests/corpus/v2_0/vocab_bridge/{seed,v0_1,v0_2_charitable}.jsonl`, gitea origin only per directory-of-origin rule, lab commit `0f0f2b3`) and ran both gates against the substrate at `github/main` HEAD `24134b6`:

| Gate | Result |
|---|---|
| Precondition (`appends ≥1 expected canonical`, threshold ≥50%) | **0/20 rows** |
| Strict NDCG@k uplift (this PR's runner) | `ON=0.6623 OFF=0.6863 uplift=-0.0240` (n=20) |

Both gates fail. The bridge regresses retrieval slightly (~2-3 NDCG points) across all 20 rows.

## Substrate root cause

Per-row debug shows the bridge appends HRR-noise tokens (`across`, `config`, `run`) instead of the designed canonicals (`vim`, `emacs`, `sqlite`, `compression`, `ruff`, etc.) — even on rows designed maximally favorably (canonicals strictly NOT in the raw query, anchored on multiple beliefs).

Inspection of `_harvest()` (`src/aelfrice/vocab_bridge.py:203-263`):

```python
def _ingest(text: str) -> None:
    for ent in extract_entities(text):
        ...
        vocab.add(low, low)          # canonical = surface
    for tok in bm25_tokenize(text):
        ...
        vocab.add(tok, tok)          # canonical = surface
```

Every surface form gets added as its own canonical. There is no cross-form mapping anywhere in the harvest. So `bridge_vec ≈ sum_t bind(t, t)` (every token bound to itself) plus HRR superposition noise. The cleanup-memory query for any `unbind(query_token, bridge_vec)` recovers the same token (self-similarity ≈ 1) — but the rewriter then drops it because it's already in the query (`vocab_bridge.py:310-311`). What gets through is HRR noise: random near-neighbors in the cleanup matrix that exceed the `1/sqrt(dim) ≈ 0.022` noise floor.

The "vocabulary-gap-recovery" claim from the spec presupposes the harvest step distinguishes surface forms from canonicals. The shipped harvest doesn't — it treats every observed token as both. There's no surface→canonical structure for HRR to recover.

## What this PR ships

The PR is correct and lands the harness regardless of substrate verdict:

- The runner produces deterministic byte-identical results across runs (verified manually on 3 invocations of the same row set).
- The unit tests pass standalone (the strict gate skips on public CI per the existing `bench_gated` autouse marker).
- The bench-gate scaffold gives operator-side a one-step `pytest -k vocab_bridge_ship_gate AELFRICE_CORPUS_ROOT=...` that fires only when both runner and labelled rows are present.

## Recommended next step (operator decision)

Two options:

1. **Close-as-not-pursued** (mirror #197 dedup R2 outcome). The bench cleared no threshold on plausible corpora; the substrate's "vocab gap recovery" claim isn't supported by `_harvest`'s self-mapping behavior. `use_vocab_bridge` stays default-OFF; the substrate stays in tree as a lane other modules can extend; the umbrella row for #433 flips to "evaluated, declined."

2. **Re-spec the bridge.** Rework `_harvest` to extract surface↔canonical *pairs* (e.g. via lemmatization, abbreviation expansion, anchor-text aliasing rules, or a pre-trained subword tokeniser) so HRR cleanup has actual structure to recover. That's a fresh design loop and a fresh PR — out of scope for #433 as filed.

Memory belief lock policy on #197 says "we don't ship a v2.x retrieval feature without positive bench delta" — option 1 is the consistent pick.

Companion to issue #474 umbrella refresh (separate small PR).

## Summary by Sourcery

Add an NDCG@k-based uplift runner and ship gate for the vocabulary bridge feature flag.

New Features:
- Introduce a VocabBridge uplift driver that runs retrieve_v2 with and without the vocabulary bridge and reports mean NDCG@k uplift.

Tests:
- Add unit tests covering empty input, skipping rows without expected_top_k, and a synthetic end-to-end row for the VocabBridge uplift driver.
- Add a bench-gated test that wires the vocab bridge corpus into the uplift runner and enforces strictly positive NDCG@k uplift before enabling the feature by default.

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive test coverage for Vocabulary Bridge feature validation, including uplift measurement and schema validation.
  * Introduced new gated test that measures NDCG@k uplift performance when Vocabulary Bridge is toggled on versus off.
  * Added unit tests verifying correct handling of edge cases and end-to-end functionality.

[Review Change Stack](https://app.coderabbit.ai/change-stack/robotrocketscience/aelfrice/pull/535)

### CODE REVIEW

**Comment by robotrocketscience** [comment]
[claim:review:Noether:2026-05-10T03:29:06Z]

**Review by robotrocketscience — COMMENTED** [review]
LGTM — rebased and FF-merging.

## Verified

- Rebased onto current `github/main` (3 atomic commits re-signed, all G).
- Local pytest on rebased tree: `tests/test_retrieve_uplift_runner.py tests/bench_gate/test_vocab_bridge_uplift.py` → 20 passed, 2 skipped (the bench-gate skips on public CI without `AELFRICE_CORPUS_ROOT`, as designed — same skip-policy contract as `#291` query-strategy and `#436` clustering gates).
- Discretion grep on diff: clean.
- Runner shape mirrors `run_doc_linker_uplift` / `run_query_strategy_uplift` — `mean_ndcg_off` / `mean_ndcg_on` / `uplift` field names preserve the branch-free formatter.
- `expected_top_k` row schema is a clean superset of the precondition row schema (`expected_canonicals` stays); rows can carry one or both, and the runner skips ones without ranking ground truth so they don't dilute the mean.
- Bench-gate test correctly assertion-pairs: `pytest.importorskip` on the runner module + skip when no row carries `expected_top_k` + `assert results.uplift > 0` for the strict gate.

## Substrate diagnosis spot-checked

Walked `src/aelfrice/vocab_bridge.py:_harvest()` on `github/main` HEAD. The PR body's claim is exactly what's there: every observed surface form is added as its own canonical (`vocab.add(low, low)` at line 223 for entities, `vocab.add(tok, tok)` at line 243 for BM25 tokens). There is no surface→canonical mapping logic in the harvest. The HRR superposition built downstream is therefore `bridge_vec ≈ Σ_t bind(t, t)` — self-bound tokens plus noise — and the rewriter's deduplication step (`vocab_bridge.py:310-311`) drops any token that's already in the query, so the rewrite path emits noise rather than canonicals. The "vocabulary-gap-recovery" claim presupposes a structure the shipped harvest doesn't build.

Per-row "appends ≥1 expected canonical" failing 0/20 plus strict NDCG@k uplift = -0.0240 is consistent with that diagnosis: the bridge isn't merely a no-op, it actively perturbs the query with low-amplitude noise, dropping retrieval slightly.

## Recommendation

Merge as evidence + harness. Keep #433 open; #536 (re-spec `_harvest`) is the right follow-up. Do **not** flip `use_vocab_bridge` to default-on at any point on the current substrate — the bench gate scaffolded here will fire correctly the moment a future substrate change re-enables the lab corpus runner, so the gate is its own forward-protection.

(Cross-session same-account auth blocks formal `--approve`; merging by local FF push per protocol.)

**Comment by robotrocketscience** [comment]
[release:review:Noether:2026-05-10T03:35:47Z]
